go/arrow/flight/record_batch_writer.go

```diff
@@ -47,7 +53,16 @@ func (f *flightPayloadWriter) WritePayload(payload ipc.Payload) error {
 	payload.SerializeBody(&f.buf)
 	f.fd.DataBody = f.buf.Bytes()
 
-	return f.w.Send(&f.fd)
+	err := f.w.Send(&f.fd)
+	if err == io.EOF {
+		// gRPC returns io.EOF if the error was generated by the server.
+		// We return a more informative error to clients that hides this gRPC idiosyncrasy.
+		// The error can be retrieved subsequently by calling RecvMsg on the stream
+		// ref: https://pkg.go.dev/google.golang.org/grpc#ClientStream
+		return ErrDataStreamClosed
+	}
```

Comment:
This is a breaking change/do we necessarily want to deviate from gRPC here?

Reply:
Perhaps not... This is a good point about being a breaking change.

Returning `io.EOF` on `Write` makes for a pretty confusing API IMO, but I suppose we're stuck with it. I'll revert this and just leave the comment in.

Reply:
Updated